src/plots/cartesian/handle_outline.js

```diff
@@ -0,0 +1,95 @@
+/**
+* Copyright 2012-2020, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var Registry = require('../../registry');
+
+
+function activateShape(gd, path, drawShapes) {
+    var element = path.node();
+    var id = +element.getAttribute('data-index');
+
+    for(var q = 0; q < gd._fullLayout.shapes.length; q++) {
+        var shapeIn = gd._fullLayout.shapes[q]._input;
+        if(q === id && shapeIn.editable) {
+            gd._fullLayout._activeShapeIndex = q;
+            break;
+        }
+    }
+
+    if(gd._fullLayout._activeShapeIndex >= 0) {
+        drawShapes(gd);
+    }
+}
+
+function deactivateShape(gd) {
+    clearOutlineControllers(gd);
+
+    var shapes = [];
+    for(var q = 0; q < gd._fullLayout.shapes.length; q++) {
+        var shapeIn = gd._fullLayout.shapes[q]._input;
+        shapes.push(shapeIn);
```

Comment:
Isn't this just recreating `gd.layout.shapes`? In which case the `_guiRelayout` call would be a noop... what am I missing?

There's also a situation where `gd._fullLayout.shapes` is not 1:1 with `gd.layout.shapes` - if you have named shapes in the template, without matching items in `shapes`. I don't know if it will do everything you need, but there's a template-aware helper for editing array items that should at least help understand the scope

https://github.com/plotly/plotly.js/blob/22dd9c1d5711575ba34c3d9f42ccf8be12398350/src/plot_api/plot_template.js#L251-L252

Reply:
> Isn't this just recreating `gd.layout.shapes`? In which case the `_guiRelayout` call would be a noop... what am I missing?
> 
> There's also a situation where `gd._fullLayout.shapes` is not 1:1 with `gd.layout.shapes` - if you have named shapes in the template, without matching items in `shapes`. I don't know if it will do everything you need, but there's a template-aware helper for editing array items that should at least help understand the scope
> 
> https://github.com/plotly/plotly.js/blob/22dd9c1d5711575ba34c3d9f42ccf8be12398350/src/plot_api/plot_template.js#L251-L252

Good call. Revised not to have any `_guiRelayout`:
https://github.com/plotly/plotly.js/blob/f7643ccb923f8efcda3a16735b59a5f3a9950dc5/src/components/shapes/draw.js#L708-L735

Reply:
> There's also a situation where `gd._fullLayout.shapes` is not 1:1 with `gd.layout.shapes` - if you have named shapes in the template, without matching items in `shapes`. I don't know if it will do everything you need, but there's a template-aware helper for editing array items that should at least help understand the scope

Very good point. Fixes are made so that the shapes from `template` don't cause major problem for new shapes as well as shapes on the `layout`. 
We'll get back to improve edits using `arrayEditor` in a separate PR.